src/equations/shallow_water_2d.jl

```diff
@@ -199,6 +197,71 @@ For details see Section 9.2.5 of the book:
   return flux
 end
 
+"""
+    boundary_condition_slip_wall(u_inner, normal_direction, direction, x, t,
+                                 surface_flux_function, equations::ShallowWaterEquations2D)
+
+Should be used together with [`StructuredMesh`](@ref).
+"""
+@inline function boundary_condition_slip_wall(u_inner, normal_direction::AbstractVector,
+                                              direction, x, t,
+                                              surface_flux_function,
+                                              equations::ShallowWaterEquations2D)
+  # flip sign of normal to make it outward pointing, then flip the sign of the normal flux back
+  # to be inward pointing on the -x and -y sides due to the orientation convention used by StructuredMesh
+  if isodd(direction)
+    boundary_flux = -boundary_condition_slip_wall(u_inner, -normal_direction,
+                                                  x, t, surface_flux_function, equations)
+  else
+    boundary_flux = boundary_condition_slip_wall(u_inner, normal_direction,
+                                                 x, t, surface_flux_function, equations)
+  end
+
+  return boundary_flux
+end
+
+"""
+    boundary_condition_slip_wall(u_inner, orientation, direction, x, t,
+                                 surface_flux_function, equations::ShallowWaterEquations2D)
+
+Should be used together with [`TreeMesh`](@ref).
+"""
+@inline function boundary_condition_slip_wall(u_inner, orientation,
+                                              direction, x, t,
+                                              surface_flux_function,
+                                              equations::ShallowWaterEquations2D)
+  # get the appropriate normal vector from the orientation
+  if orientation == 1
+    if isodd(direction)
+      normal_direction = SVector(-1, 0)
+    else
+      normal_direction = SVector(1, 0)
+    end
+  else # orientation == 2
+    if isodd(direction)
+      normal_direction = SVector(0, -1)
+    else
+      normal_direction = SVector(0, 1)
+    end
+  end
+
+  # normalize the outward pointing direction
+  normal = normal_direction / norm(normal_direction)
+
+  # compute the normal velocity
+  u_normal = normal[1] * u_inner[2] + normal[2] * u_inner[3]
+
+  # create the "external" boundary solution state
+  u_boundary = SVector(u_inner[1],
+                       u_inner[2] - 2.0 * u_normal * normal[1],
+                       u_inner[3] - 2.0 * u_normal * normal[2],
+                       u_inner[4])
+
+  # compute and return the flux using `boundary_condition_slip_wall` routine above
+  flux = surface_flux_function(u_inner, u_boundary, orientation, equations)
+
+  return flux
```

Comment:
I believe this can be simplified such that `TreeMesh` version calls the `StructuredMesh` version that then calls the original normal direction version. If you replace this code with
```julia
  # get the appropriate normal vector from the orientation
  if orientation == 1
    normal_direction = SVector(1.0, 0.0)
  else # orientation == 2
    normal_direction = SVector(0.0, 1.0)
  end

  # compute and return the flux using `boundary_condition_slip_wall` routine above
  return boundary_condition_slip_wall(u_inner, normal_direction, direction,
                                      x, t, surface_flux_function, equations)
```

This should be equivalent to the current strategy. That is, the test values should be the same but it makes the code a bit simpler.

Reply:
Yes, this should work. I will try it.

Reply:
This approach leads to the same error we had before. This is that the non conservative surface function cannot be called by when passing `normal_direction`. For `TreeMesh` we need `orientation`.